src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
```

Comment:
Given that the design above results in potentially two OS writes per call to WriteAsync, supporting what you've written above could result in confusing output to consumers. Even if there's a valid case for supporting non-deterministic writing of the inputs, each WriteAsync doesn't result in an atomic OS write, so the data could be interleaved in the output stream.

And if it's fine to require that you can't have multiple active async writes to the stream, we can change the condition to a null check which will result in less allocations. Right now when the buffer is empty, it'll still allocate a new buffer on the incoming write despite _buffer not being in use.